### ISSUE TITLE
Keycloak 18.0.0 (legacy) infinispan properties are have underscores changed

### ISSUE THREAD

**Issue by darius-m** [keycloak/keycloak#11858]
### Describe the bug

In Keycloak 18.0.0 (the legacy version) the infinispan property names have their underscores (`_`) replaced with dashes (`-`). This results in keycloak being unable to connect to the infinispan cluster.

For example, if the configuration file contains the line:
```
<property name="infinispan.client.hotrod.auth_username">keycloak</property>
```
the output will contain, depending on the keycloak version:
  * 15.1.1:
```
WARN  [org.infinispan.CONFIG] (MSC service thread 1-1) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.auth_password'. Please check your configuration. Ignoring!
```
  * 18.0.0:
```
WARN  [org.infinispan.CONFIG] (MSC service thread 1-3) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.trust-store-password'. Please check your configuration. Ignoring!
```

Using the same configuration for the infinispan subsystem, version 15.1.1 works properly, but in version 18.0.0 we get the following error:
```
org.infinispan.client.hotrod.exceptions.TransportException:: ISPN004071: Connection to <infinispan-name>:11222 was closed while waiting for response.
```

According to both [the installation guide](https://www.keycloak.org/docs/latest/server_installation/index.html#proc-configuring-remote-cache-xsite), and the official [infinispan API docs](https://docs.jboss.org/infinispan/13.0/apidocs/org/infinispan/client/hotrod/configuration/package-summary.html), the property names should contain underscores (`_`).

### Version

18.0.0

**Comment by darius-m** [comment]
This still appears to be an issue in version 18.0.2; it's probably related to the Wildfly version. Any ideas on whether it will be addressed in version 19, or if it can be fixed somehow?

**Comment by darius-m** [comment]
If it helps anyone, I currently have the following configuration (which appears to be working in 19.0.0, but I cannot confidently say that it is properly tested):
 - the cache configuration uses a `hotrod-store` instead of a `remote-store` element that defines the `rawValues` and `marshaller` properties:
```xml
<replicated-cache name="work">
    <hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
        <property name="rawValues">true</property>
        <property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
    </hotrod-store>
</replicated-cache>
```
- the remote cache container is defined outside other cache container elements (but still inside the infinispan subsystem), and contains the definitions for the infinispan authentication properties:
```xml
<remote-cache-container name="remote-container" default-remote-cluster="remote-cluster">
    <remote-clusters>
        <remote-cluster name="remote-cluster" socket-bindings="remote-cache"/>
    </remote-clusters>
    <property name="infinispan.client.hotrod.auth_username">myuser</property>
    <property name="infinispan.client.hotrod.auth_password">qwer1234!</property>
    [...]
</remote-cache-container>
```

**Comment by darius-m** [comment]
After some more testing, it appears that this approach only works inside the cluster, but does not synchronize the data in a cross datacenter deployment. This is a show-stopper, since cross-DC redundancy does not work, and user sessions are lost when accessing a node inside a different cluster.

When Infinispan debugging is enabled, the following messages are logged:
```
INFO [org.infinispan.CLUSTER] (ServerService Thread Pool -- 57) ISPN000094: Received new cluster view for channel ejb: [k_12|1] (2) [k_12, k_11]
[...]
DEBUG [org.keycloak.cluster.infinispan.CrossDCAwareCacheFactory] (ServerService Thread Pool -- 66) No configured remoteStore available. Cross-DC scenario is not used
[...]
DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserLoginFailureProviderFactory] (ServerService Thread Pool -- 66) No remote store configured for cache 'loginFailures'
DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory] (ServerService Thread Pool -- 66) No remote store configured for cache 'sessions'
DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory] (ServerService Thread Pool -- 66) No remote store configured for cache 'clientSessions'
DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory] (ServerService Thread Pool -- 66) No remote store configured for cache 'offlineSessions'
DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory] (ServerService Thread Pool -- 66) No remote store configured for cache 'offlineClientSessions'
```

This should indicate that the Keycloak nodes do join the local Infinispan clusters, but for some reason, data is not persisted between the cross-DC Infinispan nodes. From testing, this was not an issue when using `remote-store` elements for the clusters in version 15.1.1, so I am not sure how replication works if the dedicated Infinispan nodes do not handle this transparently from Keycloak.

From what I could understand from the code, Keycloak [explicitly relies](https://github.com/keycloak/keycloak/blob/main/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/InfinispanClusterProviderFactory.java#L102) on `remote-store` elements to enable cross-DC functionality, that may have been [deprecated](https://docs.wildfly.org/25/wildscribe/subsystem/infinispan/cache-container/distributed-cache/store/remote/index.html).

**Comment by darius-m** [comment]
@mposolda You are signed off as the author of this component; it would be of great help if you have any insight. Also, is there any way the new Quarkus version is affected by these issues?

The issue is currently present in Keycloak 19.0.1-legacy at least.

**Comment by mposolda** [comment]
@darius-m As mentioned in the documentation, it will be good if you use underscores. I see that in your error message for Keycloak 18, there is this:
```
WARN  [org.infinispan.CONFIG] (MSC service thread 1-3) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.trust-store-password'. Please check your configuration. Ignoring!
```
Are you sure that property `'infinispan.client.hotrod.trust-store-password` is not configured with the minus character `-` in any of the remote stores?

**Comment by darius-m** [comment]
Yes, fairly certain that the values that are set are properly using underscores.

All underscores and are replaced in the WildFly server with dashes; uppercase characters are also replaced with dash and the equivalent lowercase characters (so `rawValues` is replaced with in `raw-values`, but this issue was visible after switching to `hotrod-store` and keeping the `rawValues` property inside it, instead of the `remote-cache-container`). In testing I tried adding multiple underscores or escaping the characters with backslashes, but even `val_\\_x____y` would result in `val-\\-x---y`.

In Keycloak 15 there was no such issue, but after upgrading the 18.0.0 or 19.0.1 (and updating authentication configuration options - removing security realms mostly) it stopped working; we use Ansible to deploy the service, and the keys should be correct in the [standalone-ha.xml template](https://github.com/UPB-Admin/ansible-keycloak/blob/master/roles/keycloak/templates/standalone-ha.xml#L392), since they were not changed between versions.

**Comment by mposolda** [comment]
@tkyjovsk @lhanusov Anyone of you saw recently any issues related to underscores VS dash as specified in this GH issue?

@darius-m Just a note that we don't test with `hotrod-store` and it is not not recommended to use it. I think it probably doesn't work at all. The only allowed cross-dc setup is with the usage of `remote-store` for the caches exactly as specified in the docs: https://www.keycloak.org/docs/latest/server_installation/#proc-configuring-remote-cache-xsite (Yes, it sucks to repeat the same properties of `remote-store` for each cache and AFAIK this will be improved in the new store, but it is not there yet). Are you sure you're using latest Keycloak with `remote-store` attached to all caches as specified in the docs?

**Comment by lhanusov** [comment]
@mposolda Hi Marek, I can't help you, I do not know anything about this.

**Comment by darius-m** [comment]
It appears I was wrong about the uppercase characters (I did get an unrecognized attribute `raw-values` warning at some point, but it may have been because it was placed inside a `hotrod-store`), but as an example, after setting the following store / property:
```xml
<replicated-cache name="work" >
  <remote-store cache="work" remote-servers="remote-cache" passivation="false"
                          fetch-state="false" purge="false" preload="false" shared="true">
    <property name="rawValues">true</property>
    <property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
    <property name="infinispan.client.hotrod.auth_username__Xer_\\_y\_xyz_+=()-rr">keycloak</property>
[...]
```
the following warning is displayed:
```
WARN  [org.infinispan.CONFIG] (MSC service thread 1-2) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.auth-username--Xer-\\-y\-xyz-+=()-rr'. Please check your configuration. Ignoring!
```

**Comment by mposolda** [comment]
@lhanusov Thanks

@darius-m I am not sure why you are using `infinispan.client.hotrod.auth_username__Xer_\\_y\_xyz_+=()-rr` . Property of this name obviously doesn't exists and that's probably the reason of the warning. Do you see same warning if you use the property as specified in the docs? So something like:
```
<property name="infinispan.client.hotrod.auth_username">keycloak</property>
```

**Comment by darius-m** [comment]
No, that was just an example to showcase that it indiscriminately replaces all underscores with dashes, while leaving other characters as are. We do get `Unrecognized attribute 'infinispan.client.hotrod.auth-username'` when using the proper property, also.

**Comment by darius-m** [comment]
But, as mentioned in the documentation, in version `15.1.1` we get `Unrecognized attribute infinispan.client.hotrod.auth_username'. Please check your configuration. Ignoring!` (and similar warnings for all other properties); however, in this case the connection works properly - but this is not the case in Keycloak 18 / 19.

**Comment by darius-m** [comment]
If it helps, the upgrade to 17.0.1 appears to work with the same configuration as 18 / 19, when the normal `infinispan.client.hotrod.*` parameters are used. The only differences are that the `datasources` uses xmlns version `6.0` instead of `7.0` and `elytron` uses xmlns version `15.0` instead of `15.1` (Keycloak 19 does not work with these values either).

**Comment by mposolda** [comment]
@darius-m Thanks. So in the meantime, it is decided to remove the Wildfly distribution support for Keycloak from the latest Keycloak main. So the Keycloak 20 won't have the support for Wildfly distribution anymore. The current cross-dc support is not tested with the Keycloak quarkus distribution. So cross-dc deployments would be properly supported and tested again with the new store in the future.

So in other words, this means that this issue is pretty much becoming outdated. Keycloak team won't have time to look at it. I am closing it for now. If you figure what causes the troubles for you, you can comment here for the reference.

**Comment by artemkrim** [comment]
> If it helps anyone, I currently have the following configuration (which appears to be working in 19.0.0, but I cannot confidently say that it is properly tested):
> 
> * the cache configuration uses a `hotrod-store` instead of a `remote-store` element that defines the `rawValues` and `marshaller` properties:
> 
> ```
> <replicated-cache name="work">
>     <hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
>         <property name="rawValues">true</property>
>         <property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
>     </hotrod-store>
> </replicated-cache>
> ```
> 
> * the remote cache container is defined outside other cache container elements (but still inside the infinispan subsystem), and contains the definitions for the infinispan authentication properties:
> 
> ```
> <remote-cache-container name="remote-container" default-remote-cluster="remote-cluster">
>     <remote-clusters>
>         <remote-cluster name="remote-cluster" socket-bindings="remote-cache"/>
>     </remote-clusters>
>     <property name="infinispan.client.hotrod.auth_username">myuser</property>
>     <property name="infinispan.client.hotrod.auth_password">qwer1234!</property>
>     [...]
> </remote-cache-container>
> ```

hi, can you show full config for kk 19?

**Comment by darius-m** [comment]
Honestly there's not much to say about it - the properties in the `remote-cache-container` are the same as the ones described in the [installation guide](https://www.keycloak.org/docs/latest/server_installation/index.html#assembly-setting-up-crossdc) for the WildFly distribution and each cache has a basically identical configuration other than the cache type:

```xml
<subsystem xmlns="urn:jboss:domain:infinispan:13.0">
	<cache-container name="ejb" default-cache="dist" marshaller="PROTOSTREAM" aliases="sfsb" modules="org.wildfly.clustering.ejb.infinispan">
		<transport lock-timeout="60000"/>
		<distributed-cache name="dist">
			<locking isolation="REPEATABLE_READ"/>
			<transaction mode="BATCH"/>
			<file-store/>
		</distributed-cache>
	</cache-container>
	<cache-container name="keycloak" marshaller="JBOSS" modules="org.keycloak.keycloak-model-infinispan">
		<transport lock-timeout="60000"/>
		<local-cache name="realms">
			<heap-memory size="10000"/>
		</local-cache>
		<local-cache name="users">
			<heap-memory size="10000"/>
		</local-cache>
		<local-cache name="authorization">
			<heap-memory size="10000"/>
		</local-cache>
		<local-cache name="keys">
			<heap-memory size="1000"/>
			<expiration max-idle="3600000"/>
		</local-cache>
		<replicated-cache name="work" >
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</replicated-cache>
		<distributed-cache name="sessions" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
		<distributed-cache name="clientSessions" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
		<distributed-cache name="offlineSessions" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
		<distributed-cache name="offlineClientSessions" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
		<distributed-cache name="loginFailures" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
		<distributed-cache name="actionTokens" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<heap-memory size="-1"/>
			<expiration interval="300000" lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
		<distributed-cache name="authenticationSessions" owners="2">
			<hotrod-store remote-cache-container="remote-container" passivation="false" fetch-state="false" purge="false" preload="false" shared="true">
				<property name="rawValues">true</property>
				<property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
			</hotrod-store>
			<expiration  lifespan="900000000000000000" max-idle="-1"/>
		</distributed-cache>
	</cache-container>
	<cache-container name="server" default-cache="default" marshaller="PROTOSTREAM" aliases="singleton cluster" modules="org.wildfly.clustering.server">
		<transport lock-timeout="60000"/>
		<replicated-cache name="default">
			<transaction mode="BATCH"/>
		</replicated-cache>
	</cache-container>
	<cache-container name="web" default-cache="dist" marshaller="PROTOSTREAM" modules="org.wildfly.clustering.web.infinispan">
		<transport lock-timeout="60000"/>
		<replicated-cache name="sso">
			<locking isolation="REPEATABLE_READ"/>
			<transaction mode="BATCH"/>
		</replicated-cache>
		<distributed-cache name="dist">
			<locking isolation="REPEATABLE_READ"/>
			<transaction mode="BATCH"/>
			<file-store/>
		</distributed-cache>
		<distributed-cache name="routing"/>
	</cache-container>
	<cache-container name="hibernate" marshaller="JBOSS" modules="org.infinispan.hibernate-cache">
		<transport lock-timeout="60000"/>
		<local-cache name="local-query">
			<heap-memory size="10000"/>
			<expiration max-idle="100000"/>
		</local-cache>
		<local-cache name="pending-puts">
			<expiration max-idle="60000"/>
		</local-cache>
		<invalidation-cache name="entity">
			<transaction mode="NON_XA"/>
			<heap-memory size="10000"/>
			<expiration max-idle="100000"/>
		</invalidation-cache>
		<replicated-cache name="timestamps"/>
	</cache-container>
	<remote-cache-container name="remote-container" default-remote-cluster="remote-cluster">
		<remote-clusters>
			<remote-cluster name="remote-cluster" socket-bindings="remote-cache"/>
		</remote-clusters>
		<property name="infinispan.client.hotrod.auth_username">myuser</property>
		<property name="infinispan.client.hotrod.auth_password">qwer1234!</property>
		<property name="infinispan.client.hotrod.auth_realm">default</property>
		<property name="infinispan.client.hotrod.auth_server_name">infinispan</property>
		<property name="infinispan.client.hotrod.sasl_mechanism">SCRAM-SHA-512</property>
		<property name="infinispan.client.hotrod.trust_store_file_name">/path/to/truststore.jks</property>
		<property name="infinispan.client.hotrod.trust_store_type">JKS</property>
		<property name="infinispan.client.hotrod.trust_store_password">password</property>
	</remote-cache-container>
</subsystem>
```

However, this configuration only works inside a single cluster; the caches are not replicated between clusters, so it may simply be the case that the Keycloak servers manage to communicate locally without the Infinispan servers.

We have since switched to Quarkus, and use the Ansible template from [here](https://github.com/UPB-Admin/ansible-keycloak/blob/master/roles/keycloak/templates/cache-ispn.xml), which was based on discussion #12239. There appear to be some warnings / errors like this periodically (line broken for readability):
```
WARN  [org.infinispan.expiration.impl.ClusterExpirationManager] (expiration-thread--p5-t1) ISPN000026: Caught exception purging data container!:
org.infinispan.commons.CacheException: org.infinispan.client.hotrod.exceptions.HotRodClientException:Request for messageId=1144610 returned server error (status=0x86):
org.infinispan.util.concurrent.TimeoutException: ISPN000299: Unable to acquire lock after 0 milliseconds for key WrappedByteArray[0304090000000E\j\a\v\a\.\u\t\i\l\.\U\U\I\DBC9903F798\m85\/000000020000000C\l\e\a\s\t\S\i\g\B\i\t\s\$000000000B\m\o\s\t\S\i\g\B\i... (85 bytes)]
and requestor CommandInvocation:infinispan-server-9579:24195. Lock is held by CommandInvocation:infinispan-server-9579:14125
```

In theory, the Keycloak cache container has `<transport lock-timeout="60000"/>` set on it, so the 0 millisecond timeout may be an issue, but I am unsure if an issue or discussion should be opened for this (sorry for bothering @mposolda again for this).

**Comment by mposolda** [comment]
@darius-m Just one pointer: It seems that your XML config is different than the one described in the https://www.keycloak.org/docs/latest/server_installation/#proc-configuring-remote-cache-xsite . In the point 4 of the docs, you can see the properties as the part of the `remote-store` inside the `replicated-cache` (or `distributed-cache` elements).

The setup you use when properties are set on the `remote-cache-container` and the caches use `hotrod-store` instead of `remote-store` was not tested by Keycloak team at all.

**Comment by darius-m** [comment]
Yes, we have discussed that matter earlier in this thread, and you have confirmed that using `hotrod-store` is not supported. I did not manage to get the configuration working with `remote-store` since the property names had underscores replaced with dashes, so Keycloak failed to connect to Infinispan.

However, we have since switched to Quarkus, but there is no clear way to configure Keycloak and Infinispan to work in the Quarkus distribution, even though there were a few instances of users asking for instructions - #13926, #12239, #10694, #10521.

I think we got to a somewhat stable configuration, but I still have some questions and am not sure whether the configuration is ultimately correct. Should I open a new discussion in order to attempt to get a working configuration? If so, could I tag someone in that conversation who could help? The result should probably be published in the Keycloak guides, since the current directions provide very limited information, much less than the WildFly distribution's guide - see the [distributed caches](https://www.keycloak.org/server/caching#_securing_cache_communication) guide.

**Comment by roberth1988** [comment]
I can confirm that we have the same issues ... we are not able to use our external Infinispan anymore.
Our Setup was inspired by: https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.4/html-single/server_installation_and_configuration_guide/index

Which also states the following error is ok:
> WARN  [org.infinispan.CONFIG] (MSC service thread 1-5) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.auth_password'. Please check your configuration. Ignoring!


But now same configuration doesn't work anymore ... maybe due to "_" gets converted to "-"

`keycloak-0 keycloak 00:39:40,627 WARN  [org.infinispan.CONFIG] (MSC service thread 1-1) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.auth-username'. Please check your configuration. Ignoring!
keycloak-0 keycloak 00:39:40,627 WARN  [org.infinispan.CONFIG] (MSC service thread 1-1) ISPN000292: Unrecognized attribute 'infinispan.client.hotrod.auth-password'. Please check your configuration. Ignoring!`

**Comment by hmlnarik** [comment]
Duplicate of #14444

**Comment by martin-kanis** [comment]
If anyone is interested, here is Quarkus based configuration with external Infinispan https://github.com/keycloak/keycloak/issues/13926#issuecomment-1298406454
